frontend/src/api/stack.ts

```diff
@@ -0,0 +1,83 @@
+import { useViewStore } from '@/stores/view';
+import { World } from '@/stores/worlds';
+import { UIRequestProto } from '@/types/generated/ui';
+import { StackResponse } from '@/types/stack';
+import { useQueries, useQuery } from '@tanstack/react-query';
+import { useEffect } from 'react';
+import { useMap as useMapLibre } from 'react-map-gl/maplibre';
+import { b6 } from './client';
+
+const getStack = (request: UIRequestProto): Promise<StackResponse> => {
+    return b6.post('stack', request);
+};
+
+const stackQueryParams = (
+    request?: UIRequestProto,
+    fallback?: StackResponse
+) => {
+    return {
+        queryKey: [
+            'stack',
+            request ? 'request' : 'fallback',
+            request?.expression,
+            JSON.stringify(request?.root),
+            JSON.stringify(request?.node),
+            JSON.stringify(fallback?.proto),
+        ],
+        queryFn: () =>
+            request ? getStack(request) : Promise.resolve(fallback),
+    };
+};
+
+export const useStack = (
+    world: World['id'],
+    request?: UIRequestProto,
+    fallback?: StackResponse
+) => {
+    const query = useQuery(stackQueryParams(request, fallback));
+    const view = useViewStore((state) => state.view);
+    const viewActions = useViewStore((state) => state.actions);
+    const { [world]: maplibre } = useMapLibre();
+
+    useEffect(() => {
+        if (query.data) {
+            const newCenter =
+                query.data.proto.mapCenter?.latE7 &&
```

Comment:
It's somewhat rubbished that we'd end up not setting the map's center if your viewport happens to rest exactly on the meridian. I should probably make the field in the proto optional by adding an extra boolean indicating whether it has been set or not.